### PR TITLE
Refresh base snapshot at loading 

### PR DESCRIPTION
**Pull request by dannimad**
This approach ensures that whenever we load using pending state, serializedStateManager.snapshot and serializedStateManager.processedOps get refreshed at least once, kicking off the process at loading phase.

Once the latest snapshot is fetched in the background, we need to update the saved ops list by removing all processed ops with sequence number lower than the snapshot sequence number. These are the scenarios: 

Latest snapshot seq num < first saved op seq num: Meaning that newest snapshot is older than our first saved op. Given that we start saving ops since we attach, this scenario could only happen if we are getting the same snapshot used in previous stashed container. More specifically the only possible scenario would be snapshot seq num = first saved op seq num –1. If the snapshot seq num is lower that would mean that we are not fetching the latest snapshot available or malfunctioning on the fetchSnapshot api. For this scenario we must keep all saved ops in place since the snapshot is not changed. 

First saved op seq num <= latest snapshot seq num <= last saved op seq num:  We delete all saved ops prior to the latest snapshot seq number. 

Latest snapshot seq num > last saved op seq num: this means that processedOps array is behind from the newer snapshot. For this case, we save the latest snapshot, wait for the container to process more ops until we reach the new snapshot sequence number and update the class attributes. 

The only scenarios in which we wouldn't refresh would be if the container is closed before we catch up with latest snapshot or any error during the process.

### CODE REVIEW

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 1]
```diff
@@ -89,8 +94,24 @@ export class SerializedStateManager {
 		return this._offlineLoadEnabled;
 	}
 
+	private refreshOnCatchUpMaybe(message: ISequencedDocumentMessage) {
```
Comment (dannimad) [inline packages/loader/container-loader/src/serializedStateManager.ts:97]: This approach is taking into account that we're only going to refresh the base snapshot once at loading. If we do it more frequently, we could get rid of all new latest variables and call refreshSnapshot() from addProcessedOp instead.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 2]
```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
...
 			return { baseSnapshot, version: undefined };
 		}
 	}
 
+	/**
+	 * Fetch latest snapshot available in storage and retry indefinetely
+	 * @param supportGetSnapshotApi -
+	 */
+	private async refreshSnapshot(supportGetSnapshotApi: boolean) {
+		// This need to be resilient to failure. We can use runWithRetry
```
Comment (dannimad) [inline packages/loader/container-loader/src/serializedStateManager.ts:161]: Again, this following fetch has to retry indefinitely given that we're only going to do the refresh once. If we refresh more often we could just ignore the error and addProcessedOp could retry again.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 3]
```diff
@@ -66,9 +68,12 @@ export interface IPendingDetachedContainerState extends SnapshotWithBlobs {
 }
 
 export class SerializedStateManager {
-	private readonly processedOps: ISequencedDocumentMessage[] = [];
+	private processedOps: ISequencedDocumentMessage[] = [];
 	private snapshot: SnapshotWithBlobs | undefined;
 	private readonly mc: MonitoringContext;
+	private latestSnapshot: ISnapshotTree | undefined = undefined;
```
Comment (dannimad) [inline packages/loader/container-loader/src/serializedStateManager.ts:74]: Maybe I could keep all latest snapshot info in one object, but they are gotten at different times.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 4]
```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
...
+		} else if (snapshotSN >= firstSavedOpSN && snapshotSN <= lastSavedOpSN) {
+			// Snapshot seq num is between the first and last processed op.
+			// Remove the ops that are already part of the snapshot
+			this.processedOps.splice(0, snapshotSN - firstSavedOpSN + 1);
+			this.snapshot = { baseSnapshot: latestSnapshot, snapshotBlobs };
+		} else {
+			// snapshotSN > lastSavedOpSN
+			// Snapshot is newer than the latest processed op.
+			// We need to wait and catch up with ops to reach the snapshot's state.
+			// addProcessedOps will kick the refresh process later
```
Comment (dannimad) [inline packages/loader/container-loader/src/serializedStateManager.ts:220]: We could also avoid the refreshing from addProcessedOps by waiting here somehow.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 5]
```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi).catch((error) => {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:143]: what prevents this from being called multiple times, leading to multiple promises running?

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 6]
```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
...
 
+	/**
+	 * Fetch latest snapshot available in storage and retry indefinetely
+	 * @param supportGetSnapshotApi -
+	 */
+	private async refreshSnapshot(supportGetSnapshotApi: boolean) {
+		// This need to be resilient to failure. We can use runWithRetry
+		const { snapshot } = supportGetSnapshotApi
+			? await fetchISnapshot(this.mc, this.storageAdapter, undefined)
+			: await fetchISnapshotTree(this.mc, this.storageAdapter, undefined);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:164]: this code is duplicated a few places. maybe integrate with or create free function that handles this,

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 7]
```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
...
+		const latestSnapshot: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
+			? snapshot.snapshotTree
+			: snapshot;
+		assert(latestSnapshot !== undefined, 0x8e4 /* Snapshot should exist */);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:168]: this code is duplicated a few places. maybe integrate with or create free function that handles this,.

also, we shouldn't duplicateassert short codes. they should all be unique. however, if you move this all to a shared place, so there is only one, copying the existing code there is fine

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 3 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:74]: do it all as one object, and build the whole thing in a single place, even if multiple calls, as we never want partial state.

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 2 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:161]: don't worry about retry. its handled at a deeper layer, i don't think we need to do any here

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 5 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:143]: Nothing prevents it but we only load once and the method is only called here so there shouldn't be any multiple promises. We could add a reentrancy flag in preparation for refreshing periodically but not really needed right now imo

Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:143]: we should be defensive here such that if someone changes the usage doesn't break subtly, otherwise it could be easy to unknowingly introduce bugs.

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 8]
```diff
@@ -257,4 +288,90 @@ describe("serializedStateManager", () => {
...
+			true,
+		);
+		for (let num = 0; num < 10; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
+		const { baseSnapshot, version } = await serializedStateManager.fetchSnapshot(
+			undefined,
+			false,
+		);
+		const flushPromises = async () => new Promise(setImmediate);
```
Comment (dannimad) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:315]: I know this is not reliable. I need another way to wait for refresh to finish.

File: packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+			const attributes2 = getAndAssertAttributes(pendingState2);
+			// the base snapshot was refreshed. sequenceNumber would be the same in case we haven't
+			assert(attributes.sequenceNumber < attributes2.sequenceNumber);
+			const container3: IContainerExperimental = await loader.resolve({ url }, pendingState2);
+			const dataStore3 = (await container3.getEntryPoint()) as ITestFluidObject;
+			const string3 = await dataStore3.getSharedObject<SharedString>(stringId);
+			assert.strictEqual(string3.getText(), " a c b");
+		});
+
+		it("does not refresh on snapshot fetch failure", async () => {
```
Comment (dannimad) [inline packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts:410]: Currently the test throws the error but in reality it will retry. I  think the only thing I need to do is make the error retriable

Comment (dannimad) [reply packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts:410]: Also the title is wrong, it should refresh.

File: packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+			const attributes2 = getAndAssertAttributes(pendingState2);
+			// the base snapshot was not refreshed due to fetch failure but pending state was applied.
+			assert(attributes.sequenceNumber === attributes2.sequenceNumber);
+			const container3: IContainerExperimental = await loader.resolve({ url }, pendingState2);
+			const dataStore3 = (await container3.getEntryPoint()) as ITestFluidObject;
+			const string3 = await dataStore3.getSharedObject<SharedString>(stringId);
+			assert.strictEqual(string3.getText(), "ab");
+		});
+
+		it("refresh only while processing new ops", async () => {
```
Comment (dannimad) [inline packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts:453]: Old test when I was testing refreshing after some ops. I'll remove it.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 11]
```diff
@@ -100,29 +120,120 @@ export class SerializedStateManager {
...
+				this.snapshot = { baseSnapshot: snapshotTree, snapshotBlobs };
 			}
-			return { baseSnapshot, version };
+			return { baseSnapshot: snapshotTree, version };
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi)
+				.catch((error) => {
+					this.mc.logger.sendErrorEvent(
```
Comment (dannimad) [inline packages/loader/container-loader/src/serializedStateManager.ts:143]: A lot of logging work is still needed. Ignore this

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 12]
```diff
@@ -36,7 +36,7 @@ interface IPendingMessage {
 
 class MockStorageAdapter implements ISerializedStateManagerDocumentStorageService {
 	private readonly blobs = new Map<string, ArrayBufferLike>();
-	private readonly snapshots: ISnapshotTree[] = [];
+	private readonly snapshot: ISnapshotTree;
```
Comment (dannimad) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:39]: Simpler tests, we don't need an array

Comment (dannimad) [reply packages/loader/container-loader/src/test/serializedStateManager.spec.ts:39]: And don't pay too much attention to these tests yet. Just the comments

File: packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts [hunk 13]
```diff
@@ -0,0 +1,505 @@
...
+import { IDocumentAttributes } from "@fluidframework/protocol-definitions";
+import { wrapObjectAndOverride } from "../mocking.js";
+// eslint-disable-next-line import/no-internal-modules
+import type { IPendingContainerState } from "../../../../loader/container-loader/lib/serializedStateManager.js";
+
+const configProvider = (settings: Record<string, ConfigTypes>): IConfigProviderBase => ({
+	getRawConfig: (name: string): ConfigTypes => settings[name],
+});
+
+describeCompat(
```
Comment (dannimad) [inline packages/test/test-end-to-end-tests/src/test/refreshSerializedStateManager.spec.ts:33]: These tests are in better shape than unit tests.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 14]
```diff
@@ -65,10 +67,16 @@ export interface IPendingDetachedContainerState extends SnapshotWithBlobs {
...
+	snapshotSequenceNumber: number;
+}
+
 export class SerializedStateManager {
-	private readonly processedOps: ISequencedDocumentMessage[] = [];
+	private processedOps: ISequencedDocumentMessage[] = [];
 	private snapshot: SnapshotWithBlobs | undefined;
 	private readonly mc: MonitoringContext;
+	private latestSnapshot: SnapshotInfo | undefined;
+	private refreshSnapshotLock: boolean = false;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:79]: I think it would be better to track the promise itself, rather than a Boolean. An idea to make this even safer would be to create an immutable LazyPromise, that does the refresh `private readonly LazyPromise<SnapshotInfo>(...)`. The lazy promise won't start running until the promise is used, await, then, catch, etc

Basically all of refreshSnapshot and refreshPendingAttributes would be inlined in the lazy promises function callback. This has the added advantage that the methods are no longer callable, since they are only supported to be called once, and in a the current specific flow. 

Then in fetchSnapshot, we would start the promise by using it, like you setup the catch currently.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 15]
```diff
@@ -100,29 +120,120 @@ export class SerializedStateManager {
...
+		);
+		await this.refreshPendingAttributes(snapshotTree);
+	}
+
+	/**
+	 * Updates class snapshot and processedOps with the latest fetched snapshot
+	 *
+	 * @param baseSnapshot -
+	 */
+	private async refreshPendingAttributes(baseSnapshot: ISnapshotTree) {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:182]: i wouldn't split these functions. they are not useable independently

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 16]
```diff
@@ -100,29 +120,120 @@ export class SerializedStateManager {
...
-			return { baseSnapshot, version };
+			return { baseSnapshot: snapshotTree, version };
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi)
+				.catch((error) => {
+					this.mc.logger.sendErrorEvent(
+						{
+							eventName: "Could_Not_Refresh_Snapshot",
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:145]: this does not match the event name style we use. it should be pascal case

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 11 above)
Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:143]: wrapping the body of the refreshSnapshot in  a performance event would be a better pattern

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 15 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:182]: You mean refreshPendingAttributes and refreshSnapshot? They are being used independently. I had the logic together but in case we're behind the snapshot seq num we don't need to kick off a new refresh but just use the snapshot we have already and just wait to catch up and just refresh the class attributes

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 14 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:79]: refreshPendingAttributes could be called more than once, for the most likely scenario that we're behind the snapshot and need to wait. I still think all what you said stands for refreshSnapshot.

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 15 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:182]: If we merge both functions I guess we couldn't use the LazyPromise approach since they would need to be called twice

Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:182]: i only see one callsite for each function. where are the other calls?

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 17]
```diff
@@ -100,29 +109,80 @@ export class SerializedStateManager {
...
 	) {
 		if (this.pendingLocalState === undefined) {
-			const { snapshot, version } = supportGetSnapshotApi
-				? await fetchISnapshot(this.mc, this.storageAdapter, specifiedVersion)
-				: await fetchISnapshotTree(this.mc, this.storageAdapter, specifiedVersion);
-			const baseSnapshot: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
-				? snapshot.snapshotTree
-				: snapshot;
-			assert(baseSnapshot !== undefined, 0x8e4 /* Snapshot should exist */);
+			const { snapshotTree, version } = await getSnapshotTree(
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:112]: returning baseSnapshot rather than snapshotTree would reduce re-mapping

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 18]
```diff
@@ -171,11 +231,53 @@ export class SerializedStateManager {
...
+		storageAdapter,
+		supportGetSnapshotApi,
+		undefined,
+	);
+	const snapshotBlobs = await getBlobContentsFromTree(snapshotTree, storageAdapter);
+	const attributes: IDocumentAttributes = await getDocumentAttributes(
+		storageAdapter,
+		snapshotTree,
+	);
+	const snapshotSN = attributes.sequenceNumber;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:253]:
```suggestion
	const snapshotSequenceNumber = attributes.sequenceNumber;
```

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 19]
```diff
@@ -171,11 +231,53 @@ export class SerializedStateManager {
...
 
+async function getLatestSnapshotInfo(
+	mc: MonitoringContext,
+	storageAdapter: Pick<
+		IDocumentStorageService,
+		"getSnapshot" | "getSnapshotTree" | "getVersions" | "readBlob"
+	>,
+	supportGetSnapshotApi: boolean,
+): Promise<SnapshotInfo> {
+	const { snapshotTree } = await getSnapshotTree(
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:242]: i would like to see this all wrapped in a performance event

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 20]
```diff
@@ -119,10 +127,66 @@ export class SerializedStateManager {
...
+	 *
+	 * @param baseSnapshot -
+	 */
+	private refreshPendingAttributes() {
+		if (this.latestSnapshot === undefined) {
+			return;
+		}
+		const snapshotSequenceNumber = this.latestSnapshot?.snapshotSequenceNumber;
+		if (this.processedOps.length === 0) {
+			// weird edge case in which we don't have to do nothing to processedOps
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:154]: this comment isn't quite right. i believe its not that uncommon of a case, basically client goes offline after getting snapshot, but before it receives any remote ops. would be good to test this case.

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 21]
```diff
@@ -119,10 +127,66 @@ export class SerializedStateManager {
...
+					snapshotSequenceNumber < firstProcessedOpSequenceNumber - 1
+						? "error"
+						: "generic",
+				eventName: "OldSnapshotFetchWhileRefreshing",
+				snapshotSequenceNumber,
+				firstProcessedOpSequenceNumber,
+			});
+			this.latestSnapshot = undefined;
+		} else if (
+			snapshotSequenceNumber >= firstProcessedOpSequenceNumber &&
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:176]:
```suggestion
```
this condition is already covered by the ifs conditional, and it doesn't really matter, all the matters is the last op

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 22]
```diff
@@ -119,10 +127,66 @@ export class SerializedStateManager {
...
 			return { baseSnapshot, version: undefined };
 		}
 	}
 
+	/**
+	 * Updates class snapshot and processedOps with the latest fetched snapshot
+	 *
+	 * @param baseSnapshot -
+	 */
+	private refreshPendingAttributes() {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:148]: this should be renamed to match what it does.

Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts]: I'm bad with names :P hope you like the new one

Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts]: what you changed it to is much better. no need to change unless you want to, but here are some things i do to names things:

-  I generally `try<..>` rather than `<..>Maybe` for function names, where it doesn't always do the thing
- describe the things the functions does in a sentence, then try to simplify. usually keeping nouns and verbs
- use an online thesaurus to find synonyms to words. This can be good to get both variety, and words the fit better

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 23]
```diff
@@ -119,10 +127,61 @@ export class SerializedStateManager {
...
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot ??= (async () => {
+				this.latestSnapshot = await getLatestSnapshotInfo(
+					this.mc,
+					this.storageAdapter,
+					supportGetSnapshotApi,
+				);
+				this.updateSnapshotAndProcessedOpsMaybe();
+			})();
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:137]:
```suggestion
				this.latestSnapshot = getLatestSnapshotInfo(
					this.mc,
					this.storageAdapter,
					supportGetSnapshotApi,
				).then(()=>this.updateSnapshotAndProcessedOpsMaybe());
```
just a bit of syntax simplification

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 24]
```diff
@@ -175,7 +167,32 @@ describe("serializedStateManager", () => {
 		);
 	});
 
-	it("can get snapshot from previous local state", async () => {
+	it("can get pending local state after attach", async () => {
+		const storageAdapter = new MockStorageAdapter();
+		const serializedStateManager = new SerializedStateManager(
+			undefined,
+			logger,
+			storageAdapter,
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:175]: do we expect any calls to be made on the storage adapter is this case? if not, using something like the failProxy from packages\loader\container-loader\src\test\loader.spec.ts is a good way to validate that

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 23 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:138]: I believe they're not entirely equivalent since latestSnapshot is no longer a SnapshotInfo object. If this is needed, then there are more things in the code that need to be changed

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 25]
```diff
@@ -119,10 +128,64 @@ export class SerializedStateManager {
...
 	}
 
+	/**
+	 * Updates class snapshot and processedOps if we have a new snapshot and it's among processedOps range.
+	 */
+	private updateSnapshotAndProcessedOpsMaybe() {
+		if (this.latestSnapshot === undefined) {
+			return;
+		}
+		this.snapshotRefreshed?.();
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:151]: shouldn't this be called around 185 such that its called when the snapshot is actually refreshed?

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 26]
```diff
@@ -119,10 +128,64 @@ export class SerializedStateManager {
...
+		const snapshotSequenceNumber = this.latestSnapshot?.snapshotSequenceNumber;
+		if (this.processedOps.length === 0) {
+			// case in which we don't have to do nothing to processedOps
+			// since we don't have any.
+			this.snapshot = this.latestSnapshot;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:156]: this seems wrong. its possible there are no op in process ops, but there are still ops that must be process between the current snapshot and the downloaded snapshot. it should be possible to write a test to simulate this. i think the right behavior is to do nothing.

Comment (anthony-murphy) [reply packages/loader/container-loader/src/serializedStateManager.ts:156]: The scenario is, client downloads snapshot and not ops are processed. the client goes offline, and creates changes. the pending state is retrieved. the saved ops would be empty. when the pending state is loaded processed would be empty, but there could still be ops created remotely between the current snapshot, and the new snapshot

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 27]
```diff
@@ -20,90 +21,105 @@ import {
...
 	IDocumentStorageService,
 	"getSnapshot" | "getSnapshotTree" | "getVersions" | "readBlob"
 >;
 
-interface IPendingMessage {
-	type: "message";
-	referenceSequenceNumber: number;
-	content: string;
-}
+const failProxy = <T extends object>() => {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:32]: rather than copy, please move this a new file in the test folder, and reuse from there

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 28]
```diff
@@ -20,90 +21,105 @@ import {
...
+		_count: number,
+		_scenarioName?: string | undefined,
+		_fetchSource?: FetchSource | undefined,
 	): Promise<IVersion[]> {
-		return [{ id: "test", treeId: "test" }];
+		assert.ok(this.snapshot.id);
+		return [{ id: this.snapshot.id, treeId: this.snapshot.id }];
 	}
 	public async readBlob(id: string): Promise<ArrayBufferLike> {
 		return this.blobs.get(id) as ArrayBufferLike;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:104]: this should throw if the blob does not exist in the blobs map

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 29]
```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
...
+	});
+
+	it("refresh snapshot when snapshot sequence number is above processed ops", async () => {
+		const pending: IPendingContainerState = {
+			...pendingLocalState,
+			baseSnapshot: { ...snapshot, id: "fromPending" },
+		};
+		const storageAdapter = new MockStorageAdapter();
+		const deferred = new Deferred<void>();
+		const cb = () => {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:496]: can you improve the names here. its hard to follow the test as the current names don't give any hint of what they will be used for

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 30]
```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
...
+		}
+
+		const snapshotSequenceNumber = 22;
+		storageAdapter.uploadSummary(snapshotSequenceNumber);
+
+		await serializedStateManager.fetchSnapshot(undefined, false);
+		const processedOpsSize2 = 10;
+		for (let num = 0; num < processedOpsSize2; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:518]: this is really hard to follow. i think changing generateSavedOp to take a sequence number would make it easier, and then the test would manage the sequence numbers explicitly, rather than implicitly via counts
```suggestion
		let seq = 0;
		while(seq <20){
		     serializedStateManager.addProcessedOp(generateSavedOp(seq++));
		}
		storageAdapter.uploadSummary(seq);

		await serializedStateManager.fetchSnapshot(undefined, false);
		
		while(seq <10){
		     serializedStateManager.addProcessedOp(generateSavedOp(seq++));
		}
```

Comment (anthony-murphy) [reply packages/loader/container-loader/src/test/serializedStateManager.spec.ts:518]: this has the added advantage of getting ride of shared state, which makes code make more complicated to maintain.

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 31]
```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
...
-		assert.strictEqual(parsed.baseSnapshot.id, "SnapshotId");
+		// We keep using the pending snapshot since it is the same based on its sequence number 0
+		assert.strictEqual(parsed.baseSnapshot.id, "fromPending");
 		const attributes = getAttributesFromPendingState(parsed);
 		assert.strictEqual(attributes.sequenceNumber, 0);
 		assert.strictEqual(attributes.minimumSequenceNumber, 0);
 	});
+
+	it("snapshot is older than first processed op", async () => {
+		const pending: IPendingContainerState = {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:318]: i would like to see some tests where the pending state contains saved ops before the refresh

File: packages/loader/container-loader/src/test/serializedStateManager.spec.ts [hunk 32]
```diff
@@ -206,55 +237,309 @@ describe("serializedStateManager", () => {
...
+			true,
+		);
+
+		const processedOpsSize = 20;
+		for (let num = 0; num < processedOpsSize; ++num) {
+			serializedStateManager.addProcessedOp(generateSavedOp());
+		}
+
+		const snapshotSequenceNumber = 11;
+		storageAdapter.uploadSummary(snapshotSequenceNumber);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/test/serializedStateManager.spec.ts:382]: i would like to see some tests where the background refresh fails

File: packages/loader/container-loader/src/serializedStateManager.ts [hunk 33]
```diff
@@ -171,11 +234,84 @@ export class SerializedStateManager {
...
+		IDocumentStorageService,
+		"getSnapshot" | "getSnapshotTree" | "getVersions" | "readBlob"
+	>,
+	supportGetSnapshotApi: boolean,
+): Promise<SnapshotInfo> {
+	return PerformanceEvent.timedExecAsync(
+		mc.logger,
+		{ eventName: "GetLatestSnapshotInfo" },
+		async () => {
+			const { baseSnapshot } = await getSnapshotTree(
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/serializedStateManager.ts:257]: add a test a unit test where one of these calls throws an exception

File: packages/loader/container-loader/src/serializedStateManager.ts (hunk 25 above)
Comment (dannimad) [reply packages/loader/container-loader/src/serializedStateManager.ts:151]: It depends on what we want tests to check and when. I was more interested in knowing when the new snapshot was fetched than when we do the this.snapshot update. That allowed me to track the catch up case a bit more clearly. I'll reconsider where I put this and how I call it after updating some tests

**Review by anthony-murphy — APPROVED** [review]
